### PR TITLE
Kill sstable::storage::get_stats() to help S3 provide accurate SSTable component stats

### PR DESCRIPTION
**Pull request by raphaelsc**
S3 wasn't providing filter size and accurate size for all SSTable components on disk.

First, filter size is provided by taking advantage that its in-memory representation is roughly the same as on-disk one.

Second, size for all components is provided by piggybacking on sstable parser and writer, so no longer a need to do a separate additional step after Scylla have either parsed or written all components.

Finally, sstable::storage::get_stats() is killed, so the burden is no longer pushed on the storage type implementation.

Refs #13649.

### CODE REVIEW

File: sstables/sstables.cc [hunk 1]
```diff
@@ -1097,28 +1097,32 @@ future<> sstable::read_simple(T& component, const io_priority_class& pc) {
...
-        });
-    }).then_wrapped([file_path] (future<> f) {
         try {
-            f.get();
-        } catch (std::system_error& e) {
+
+        file fi = co_await new_sstable_component_file(_read_error_handler, Type, open_flags::ro);
+        uint64_t size = co_await fi.size();
+
+        auto r = make_lw_shared<file_random_access_reader>(std::move(fi), size, sstable_buffer_size);
```
Comment (xemul) [inline sstables/sstables.cc:1105]: It looks like `r` doesn't need to be lw_shared_ptr -- it doesn't outsurvive this function, and `parse()` accepts it by reference, now by lw-pointer

Comment (raphaelsc) [reply sstables/sstables.cc:1105]: you're right.

**Comment by raphaelsc** [comment]
> Also comment to last patch -- shouldn't the `rewrite_statistics` also be updated? It may change the size of metadata

rewrite_statistics is a hack for mutating an existing Statistics which was already loaded in memory, therefore no update is needed.

**Comment by raphaelsc** [comment]
> > Also comment to last patch -- shouldn't the `rewrite_statistics` also be updated? It may change the size of metadata
> 
> rewrite_statistics is a hack for mutating an existing Statistics which was already loaded in memory, therefore no update is needed.

Ah you're referring to a possible change of metadata size, but I don't think that will ever happen in practice. SSTables are immutable and entries that we might need to change like sstable_level won't impact the size.

I am even considering to get rid of rewrite_statistics() in the near future. Goes against the principle of immutabitility.

**Comment by xemul** [comment]
> Ah you're referring to a possible change of metadata size ...

Not exactly. I'm worried that writing of some components was missed in the patch and we end up seeing (slightly) wrong data. Maybe there's a safer way to collect write- and read-time file sizes rather than putting `_size += writer.offset()` here and there?

In particular -- how about `write_toc()`? It doesn't use `do_write_simple` so should be patched together with `write_crc()` and `write_digest()`, shouldn't it? Same for `read_toc()`, it doesn't go through patched `read_simple()`. We're not losing much with TOC, but still.

File: sstables/sstables.cc [hunk 2]
```diff
@@ -1097,28 +1097,32 @@ future<> sstable::read_simple(T& component, const io_priority_class& pc) {
...
+        co_await r.close();
+
+        if (eptr) {
+            std::rethrow_exception(std::move(eptr));
+        }
+
+        }  catch (std::system_error& e) {
             if (e.code() == std::error_code(ENOENT, std::system_category())) {
-                return make_exception_future<>(malformed_sstable_exception(file_path + ": file not found"));
+                throw malformed_sstable_exception(file_path + ": file not found");
```
Comment (xemul) [inline sstables/sstables.cc]: Since now it's a coroutine, maybe the `co_await coroutine::return_exception()` should be used?

**Comment by raphaelsc** [comment]
> > Ah you're referring to a possible change of metadata size ...
> 
> Not exactly. I'm worried that writing of some components was missed in the patch and we end up seeing (slightly) wrong data. Maybe there's a safer way to collect write- and read-time file sizes rather than putting `_size += writer.offset()` here and there?
> 

Initially I thought that writing and parsing of all components, but data and index, would go through write_simple() and read_simple(). But then I started realizing that digest and crc took a different approach.

do_write_simple has the following interface:
```
        do_write_simple(component_type type, const io_priority_class& pc,
            noncopyable_function<void (version_types version, file_writer& writer)> write_component)

```

so we can potentially patch everything to go through do_write_simple().

And do the same for all parsing, with a new do_read_simple().

**Comment by xemul** [comment]
> Makes sense to you?

Yes, one API for simple components makes perfect sense :+1:

**Comment by raphaelsc** [comment]
v3:
- write and parse all simple components through do_write_simple and do_read_simple, respectively
- add test to ensure correctness

File: sstables/sstables.cc (hunk 2 above)
Comment (xemul) [reply sstables/sstables.cc]: @raphaelsc , this suggestion is still relevant from my side :)

**Comment by xemul** [comment]
I've a doubt about replacement of `return make_exception_future<>()` with `throw` (rather than with `co_await return_exception()`. Other than that LGTM, awesome improvement!

File: sstables/sstables.cc (hunk 2 above)
Comment (raphaelsc) [reply sstables/sstables.cc]: > @raphaelsc , this suggestion is still relevant from my side :)

IIRC, we cannot use co_await inside the catch statement. So we'd need all that verbose idiom of saving exception and maybe rethrowing it post catch. I took a look at raft/ and noticed they're throwing directly in coroutine frames, inside the catch statements... The only difference, according to the doc, is that the coroutine::exception reduces one hop by propagating the exception into the coroutine directly, but for the system_error and malformed_sstable_exception, which are very rare anyway, we can just use throw for gaining readability. What do you think?

**Comment by raphaelsc** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/914/

That's https://github.com/scylladb/scylladb/issues/13211

**Comment by raphaelsc** [comment]
> I've a doubt about replacement of `return make_exception_future<>()` with `throw` (rather than with `co_await return_exception()`. Other than that LGTM, awesome improvement!

Thanks @xemul. Replied to your concern here https://github.com/scylladb/scylladb/pull/13682#discussion_r1179295461. If you agree with my reasoning, I will just send my next rebased version.

**Comment by raphaelsc** [comment]
v4:
rebased
resolved conflict with @xemul's patch to remove unused exception_ptr.

**Comment by raphaelsc** [comment]
> CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/928/

@scylladb/scylla-maint ping for merge

**Pull request merged**